hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/MixTestAnnotationsCondition.java

```diff
@@ -51,18 +63,45 @@ public MixTestAnnotationsCondition() {
 
     @Override
     public void check(JavaClass item, ConditionEvents events) {
-        boolean hasJUnit4Annotation = item.getMethods().stream()
-                .anyMatch(method -> JUNIT_4_ANNOTATION_CLASSES.stream()
-                        .anyMatch(method::isAnnotatedWith));
+        boolean hasJUnit4Elements = hasAnyJUnit4Annotations(item) || hasJUnit4Assertions(item);
+        boolean hasJUnit5Elements = hasAnyJUnit5Annotations(item) || hasJUnit5Assertions(item);
+
+        if (hasJUnit4Elements && hasJUnit5Elements) {
+            String message = String.format("Class %s mixes JUnit 4 and JUnit 5 elements.", item.getName());
+            events.add(SimpleConditionEvent.violated(item, message));
+        }
+    }
+
+    private boolean hasJUnit5Assertions(JavaClass item) {
+        Set<JavaMethodCall> methodCalls = item.getMethodCallsFromSelf();
+        return methodCalls.stream().anyMatch(call -> call.getTarget().getFullName().contains("org.junit.jupiter.api.Assertions"));
+    }
+    private boolean hasJUnit4Assertions(JavaClass item) {
+        Set<JavaMethodCall> methodCalls = item.getMethodCallsFromSelf();
+        return methodCalls.stream().anyMatch(call -> call.getTarget().getFullName().contains("org.junit.Assert"));
+    }
```

Comment:
nit - it's possible to centralise the logic into a single method that these can delegate to with different `fullName` parameters